mirrord/cli/src/port_forward.rs

```diff
@@ -442,12 +420,255 @@ impl PortForwarder {
+    waiting_for_pong: bool,
+    ping_pong_timeout: Instant,
+}
+
+impl ReversePortForwarder {
+    pub(crate) async fn new(
+        agent_connection: AgentConnection,
+        mappings: HashMap<RemotePort, LocalPort>,
+        network_config: IncomingConfig,
+    ) -> Result<Self, PortForwardError> {
+        // setup IncomingProxy
+        let mut background_tasks: BackgroundTasks<MainTaskId, ProxyMessage, IntProxyError> =
+            Default::default();
+        let incoming =
+            background_tasks.register(IncomingProxy::default(), MainTaskId::IncomingProxy, 512);
+        // construct IncomingMode from config file
+        let incoming_mode = IncomingMode::new(&network_config);
+        for (i, (&remote, &local)) in mappings.iter().enumerate() {
+            // send subscription to incoming proxy
+            let subscription = incoming_mode.subscription(remote);
+            let message_id = i as u64;
+            let layer_id = LayerId(1);
+            let req = IncomingRequest::PortSubscribe(PortSubscribe {
+                listening_on: format!("127.0.0.1:{local}")
+                    .parse()
+                    .expect("Error parsing socket address"),
+                subscription,
+            });
+            incoming
+                .send(IncomingProxyMessage::LayerRequest(
+                    message_id, layer_id, req,
+                ))
+                .await;
+        }
+
+        Ok(Self {
+            incoming_mode,
+            agent_connection,
+            mappings,
+            background_tasks,
+            incoming_proxy: incoming,
+            waiting_for_pong: false,
+            ping_pong_timeout: Instant::now(),
+        })
+    }
+
+    pub(crate) async fn run(&mut self) -> Result<(), PortForwardError> {
+        // setup agent connection
+        self.agent_connection
+            .sender
+            .send(ClientMessage::SwitchProtocolVersion(
+                mirrord_protocol::VERSION.clone(),
+            ))
+            .await?;
+        match self.agent_connection.receiver.recv().await {
+            Some(DaemonMessage::SwitchProtocolVersionResponse(version))
+                if CLIENT_READY_FOR_LOGS.matches(&version) =>
+            {
+                self.agent_connection
+                    .sender
+                    .send(ClientMessage::ReadyForLogs)
+                    .await?;
+            }
+            _ => return Err(PortForwardError::AgentConnectionFailed),
+        }
+
+        for remote_port in self.mappings.keys() {
+            let subscription = self.incoming_mode.subscription(*remote_port);
+            let msg = match subscription {
+                PortSubscription::Steal(steal_type) => {
+                    ClientMessage::TcpSteal(LayerTcpSteal::PortSubscribe(steal_type))
+                }
+                PortSubscription::Mirror(_) => {
+                    ClientMessage::Tcp(LayerTcp::PortSubscribe(*remote_port))
+                }
+            };
```

Comment:
Can this existing method be used here?

```suggestion
            let msg = subscription.agent_subscribe();
```